server/web_ui/shared/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+/// (kopid, status, json, headers)
+pub async fn do_request(
+    uri: &str,
+    method: RequestMethod,
+    body: Option<JsValue>,
+) -> Result<(Option<String>, u16, JsValue, Headers), FetchError> {
+    let mut opts = RequestInit::new();
+    opts.method(&method.to_string());
+    opts.mode(RequestMode::SameOrigin);
+    opts.credentials(web_sys::RequestCredentials::SameOrigin);
+
+    if let Some(body) = body {
+        #[cfg(debug_assertions)]
+        if method == RequestMethod::GET {
+            gloo::console::debug!("This seems odd, you've supplied a body with a GET request?")
+        }
+        opts.body(Some(&body));
+    }
+
+    let request = Request::new_with_str_and_init(uri, &opts)?;
+    request
+        .headers()
+        .set(CONTENT_TYPE, APPLICATION_JSON)
+        .expect_throw("failed to set content-type header");
+
+    if let Some(sessionid) = pop_auth_session_id() {
+        request
+            .headers()
+            .set(KSESSIONID, &sessionid)
+            .expect_throw(&format!("failed to set {} header", KSESSIONID));
+    }
+
+    if let Some(bearer_token) = get_bearer_token() {
+        request
+            .headers()
+            .set("authorization", &bearer_token)
+            .expect_throw("failed to set authorization header");
+    }
+
+    let window = utils::window();
+    let resp_value = JsFuture::from(window.fetch_with_request(&request)).await?;
+    let resp: Response = resp_value.dyn_into().expect_throw("Invalid response type");
+    let status = resp.status();
+    let headers: Headers = resp.headers();
+
+    if let Some(sessionid) = headers.get(KSESSIONID).ok().flatten() {
+        push_auth_session_id(sessionid);
+    }
+
+    let kopid = headers.get(KOPID).ok().flatten();
+
+    let body = match resp.json() {
+        Ok(json_future) => match JsFuture::from(json_future).await {
+            Ok(body) => body,
+            Err(e) => {
+                let e_msg = format!("future json error -> {:?}", e);
+                console::error!(e_msg.as_str());
+                JsValue::NULL
+            }
+        },
+        Err(e) => {
+            let e_msg = format!("response json error -> {:?}", e);
+            console::error!(e_msg.as_str());
+            JsValue::NULL
+        }
+    };
+
+    Ok((kopid, status, body, headers))
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum RequestMethod {
+    GET,
+    POST,
+    PUT,
+}
+
+impl ToString for RequestMethod {
+    fn to_string(&self) -> String {
+        match self {
+            RequestMethod::PUT => "PUT".to_string(),
+            RequestMethod::POST => "POST".to_string(),
+            RequestMethod::GET => "GET".to_string(),
+        }
+    }
+}
+
+/// creates the "Kanidm is alpha" banner
+pub fn alpha_warning_banner() -> Html {
+    html!(
+        <div class={CSS_ALERT_WARNING} role="alert">
+        {"🦀 Kanidm is still in early Alpha, this interface is a placeholder! "}
```

Comment:
I wonder if we should yeet this at this point, we are pretty mature at this point.

Reply:
It's only on the admin things, and they're still *very* alpha.